### PR TITLE
pkg/idtools: mkdirAs(): fix infinite loops and repeated "chown"

### PR DESCRIPTION
**Pull request by thaJeztah**
- relates to https://github.com/moby/moby/pull/44196#discussion_r980996184

This fixes an inifinite loop in mkdirAs(), used by `MkdirAllAndChown`, `MkdirAndChown`, and `MkdirAllAndChownNew`, as well as directories being chown'd multiple times when relative paths are used.

The for loop in this function was incorrectly assuming that;

1. `filepath.Dir()` would always return the parent directory of any given path
2. traversing any given path to ultimately result in "/"

While this is correct for absolute and "cleaned" paths, both assumptions are incorrect in some variations of "path";

1. for paths with a trailing path-separator ("some/path/"), or dot ("."), `filepath.Dir()` considers the (implicit) "." to be a location _within_ the directory, and returns "some/path" as ("parent") directory. This resulted in the path itself to be included _twice_ in the list of paths to chown.
2. for relative paths ("./some-path", "../some-path"), "traversing" the path would never end in "/", causing the for loop to run indefinitely:

    ```go
    // walk back to "/" looking for directories which do not exist
    // and add them to the paths array for chown after creation
    dirPath := path
    for {
        dirPath = filepath.Dir(dirPath)
        if dirPath == "/" {
            break
        }
        if _, err := os.Stat(dirPath); err != nil && os.IsNotExist(err) {
            paths = append(paths, dirPath)
        }
    }
    ```

A _partial_ mitigation for this would be to use `filepath.Clean()` before using the path (while `filepath.Dir()` _does_ call `filepath.Clean()`, it only does so _after_ some processing, so only cleans the result). Doing so would prevent the double chown from happening, but would not prevent the "final" path to be "." or ".." (in the relative path case), still causing an infinite loop, or additional checks for "." / ".." to be needed.

| path           | filepath.Dir(path) | filepath.Dir(filepath.Clean(path)) |
|----------------|--------------------|------------------------------------|
| some-path      | .                  | .                                  |
| ./some-path    | .                  | .                                  |
| ../some-path   | ..                 | ..                                 |
| some/path/     | some/path          | some                               |
| ./some/path/   | some/path          | some                               |
| ../some/path/  | ../some/path       | ../some                            |
| some/path/.    | some/path          | some                               |
| ./some/path/.  | some/path          | some                               |
| ../some/path/. | ../some/path       | ../some                            |
| /some/path/    | /some/path         | /some                              |
| /some/path/.   | /some/path         | /some                              |

Instead, this patch adds a `filepath.Abs()` to the function, so make sure that paths are both cleaned, and not resulting in an infinite loop.

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: pkg/idtools/idtools_unix_test.go [hunk 1]
```diff
@@ -127,6 +128,86 @@ func TestMkdirAllAndChownNew(t *testing.T) {
...
+			out: []string{"dir10", "dir10/subdir10"},
+		},
+	}
+
+	// Set the current working directory to the temp-dir, as we're
+	// testing relative paths. This may break when we run tests in
+	// parallel.
+	dirName := t.TempDir()
+	err := os.Chdir(dirName)
+	assert.NilError(t, err)
```
Comment (thaJeztah) [inline pkg/idtools/idtools_unix_test.go:185]: Not sure what a good alternative would be for this 🤔

**Comment by thaJeztah** [comment]
@corhere @neersighted ptal 🤗

File: pkg/idtools/idtools_unix_test.go [hunk 2]
```diff
@@ -127,6 +128,86 @@ func TestMkdirAllAndChownNew(t *testing.T) {
...
+
+	const expectedUIDGID = 101
+
+	for _, tc := range tests {
+		tc := tc
+		t.Run(tc.in, func(t *testing.T) {
+			for _, p := range tc.out {
+				_, err := os.Stat(p)
+				assert.Assert(t, errors.Is(err, os.ErrNotExist))
```
Comment (neersighted) [inline pkg/idtools/idtools_unix_test.go:194]: I think converting most of these `Assert`s to `Check`s would make debugging this easier if it has to be debugged in the future.

(this includes use of `assert.Equal` which could be `assert.Check` combined with `cmp.Equal`)

File: pkg/idtools/idtools_unix_test.go [hunk 3]
```diff
@@ -127,6 +128,86 @@ func TestMkdirAllAndChownNew(t *testing.T) {
...
+
+			err := MkdirAllAndChownNew(tc.in, 0755, Identity{UID: expectedUIDGID, GID: expectedUIDGID})
+			assert.NilError(t, err)
+
+			for _, p := range tc.out {
+				s := &unix.Stat_t{}
+				err = unix.Stat(p, s)
+				assert.NilError(t, err)
+				assert.Equal(t, uint64(s.Uid), uint64(expectedUIDGID))
+				assert.Equal(t, uint64(s.Gid), uint64(expectedUIDGID))
```
[the remark this answers is not in the document]
Comment (thaJeztah) [reply pkg/idtools/idtools_unix_test.go:205]: > Please restore the CWD after the test so as not to pollute the state for other tests. 

Good one! I was actually thinking of doing that, but didn't...

File: pkg/idtools/idtools_unix_test.go (hunk 2 above)
Comment (thaJeztah) [reply pkg/idtools/idtools_unix_test.go:194]: Ah, yes, @corhere commented something similar above; I'll update

**Comment by thaJeztah** [comment]
Updated; ptal 👍

File: pkg/idtools/idtools_unix_test.go [hunk 4]
```diff
@@ -127,6 +128,99 @@ func TestMkdirAllAndChownNew(t *testing.T) {
...
+
+// Change the current working directory for the duration of the test. This may
+// break if tests are run in parallel.
+func setWorkingDirectory(t *testing.T, dir string) {
+	t.Helper()
+	cwd, err := os.Getwd()
+	assert.NilError(t, err)
+	t.Cleanup(func() {
+		if err := os.Chdir(cwd); err != nil {
+			t.Log(err)
```
Comment (corhere) [inline pkg/idtools/idtools_unix_test.go:217]: Fail the test which pollutes the environment and can't revert the change. Don't let some other (innocent) test take the fall. Logged messages in a passing test are hidden by default, and even when emitted could be overlooked when trying to debug a mysterious "impossible" flake of another test.

**Review by neersighted — APPROVED** [review]

**Review by corhere — APPROVED** [review]
LGTM!
